src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs

```diff
@@ -318,26 +320,72 @@ static MsQuicApi()
             return secConfig;
         }
 
-        public unsafe IntPtr SessionOpen(byte[] alpn)
+        public unsafe IntPtr SessionOpen(List<SslApplicationProtocol> alpnProtocols)
         {
-            IntPtr sessionPtr = IntPtr.Zero;
-            uint status;
+            if (alpnProtocols.Count == 1)
+            {
+                return SessionOpen(alpnProtocols[0]);
+            }
+
+            var memoryHandles = ArrayPool<MemoryHandle>.Shared.Rent(alpnProtocols.Count);
+            var quicBuffers = ArrayPool<MsQuicNativeMethods.QuicBuffer>.Shared.Rent(alpnProtocols.Count);
 
-            fixed (byte* pAlpn = alpn)
+            try
             {
-                var alpnBuffer = new MsQuicNativeMethods.QuicBuffer
+                for (int i = 0; i < alpnProtocols.Count; ++i)
                 {
-                    Length = (uint)alpn.Length,
-                    Buffer = pAlpn
-                };
-
-                status = SessionOpenDelegate(
-                    _registrationContext,
-                    &alpnBuffer,
-                    1,
-                    IntPtr.Zero,
-                    ref sessionPtr);
+                    ReadOnlyMemory<byte> alpnProtocol = alpnProtocols[i].Protocol;
+                    MemoryHandle h = alpnProtocol.Pin();
+
+                    memoryHandles[i] = h;
+                    quicBuffers[i].Buffer = (byte*)h.Pointer;
+                    quicBuffers[i].Length = (uint)alpnProtocol.Length;
+                }
+
+                IntPtr session;
+
+                fixed (MsQuicNativeMethods.QuicBuffer* pQuicBuffers = quicBuffers)
+                {
+                    session = SessionOpen(pQuicBuffers, (uint)alpnProtocols.Count);
+                }
+
+                ArrayPool<MsQuicNativeMethods.QuicBuffer>.Shared.Return(quicBuffers);
```

Comment:
Shouldn't this be in the finally block? (Sorry, guess I missed this in the original review)

Reply:
Elsewhere we've avoided this to prevent accidentally using buffer in multiple places, but since this is P/Invoke yea it probably could be done.

Regardless, the array will just be GCed on the error path, so I don't think this is blocking?

Reply:
Ok